internal/http/services/owncloud/ocdav/copy.go

```diff
@@ -32,56 +33,47 @@ import (
 	"github.com/cs3org/reva/internal/http/services/datagateway"
 	"github.com/cs3org/reva/pkg/appctx"
 	"github.com/cs3org/reva/pkg/rhttp"
+	"github.com/rs/zerolog"
 	"go.opencensus.io/trace"
 )
 
-func (s *svc) handleCopy(w http.ResponseWriter, r *http.Request, ns string) {
+type copy struct {
+	sourceInfo  *provider.ResourceInfo
+	destination *provider.Reference
+	depth       string
+	successCode int
+}
+
+type intermediateDirRefFunc func() (*provider.Reference, *rpc.Status, error)
+
+func (s *svc) handlePathCopy(w http.ResponseWriter, r *http.Request, ns string) {
 	ctx := r.Context()
 	ctx, span := trace.StartSpan(ctx, "head")
 	defer span.End()
 
 	src := path.Join(ns, r.URL.Path)
-	dstHeader := r.Header.Get("Destination")
-	overwrite := r.Header.Get("Overwrite")
-	depth := r.Header.Get("Depth")
-	if depth == "" {
-		depth = "infinity"
-	}
-
-	dst, err := extractDestination(dstHeader, r.Context().Value(ctxKeyBaseURI).(string))
+	dst, err := extractDestination(r)
 	if err != nil {
 		w.WriteHeader(http.StatusBadRequest)
 		return
 	}
 	dst = path.Join(ns, dst)
 
 	sublog := appctx.GetLogger(ctx).With().Str("src", src).Str("dst", dst).Logger()
-	sublog.Debug().Str("overwrite", overwrite).Str("depth", depth).Msg("copy")
 
-	overwrite = strings.ToUpper(overwrite)
-	if overwrite == "" {
-		overwrite = "T"
-	}
+	srcRef := &provider.Reference{Path: src}
 
-	if overwrite != "T" && overwrite != "F" {
-		w.WriteHeader(http.StatusBadRequest)
-		m := fmt.Sprintf("Overwrite header is set to incorrect value %v", overwrite)
-		b, err := Marshal(exception{
-			code:    SabredavBadRequest,
-			message: m,
-		})
-		HandleWebdavError(&sublog, w, b, err)
-		return
+	// check dst exists
+	dstRef := &provider.Reference{Path: dst}
+
+	intermediateDirRefFunc := func() (*provider.Reference, *rpc.Status, error) {
+		intermediateDir := path.Dir(dst)
+		ref := &provider.Reference{Path: intermediateDir}
+		return ref, &rpc.Status{Code: rpc.Code_CODE_OK}, nil
 	}
 
-	if depth != "infinity" && depth != "0" {
-		w.WriteHeader(http.StatusBadRequest)
-		m := fmt.Sprintf("Depth header is set to incorrect value %v", depth)
-		b, err := Marshal(exception{
-			code:    SabredavBadRequest,
-			message: m,
-		})
-		HandleWebdavError(&sublog, w, b, err)
+	cp := s.prepareCopy(ctx, w, r, srcRef, dstRef, intermediateDirRefFunc, &sublog)
```

Comment:
Can we call `prepareCopy` from `executePathCopy` (maybe rename it)?

Reply:
Not really, executePathCopy is a recursive function and we would need to introduce some condition to only execute prepareCopy once in executePathCopy.

Reply:
Also later there will be a `executeSpacesCopy` which implements spaces specific stuff, that is why `prepareCopy` is a function on its own.